glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,375 @@
+    ///   monotonically increasing timer (this value is obtained on the
+    ///   platform-specific side).
+    /// - `extra`: Extra data values, mapping strings to strings.
+    pub fn record(
+        &self,
+        glean: &Glean,
+        meta: &CommonMetricData,
+        timestamp: u64,
+        extra: Option<HashMap<String, String>>,
+    ) {
+        let event = RecordedEventData {
+            timestamp,
+            category: meta.category.to_string(),
+            name: meta.name.to_string(),
+            extra,
+        };
+        let event_json = event.serialize().to_string();
+        let mut stores_to_send: Vec<&str> = Vec::new();
+
+        {
+            let mut db = self.event_stores.write().unwrap();
+            for store_name in meta.send_in_pings.iter() {
+                let store = db
+                    .entry(store_name.to_string())
+                    .or_insert_with(|| Vec::new());
+                store.push(event.clone());
+                self.write_event_to_disk(store_name, &event_json);
+                if store.len() == glean.get_max_events() {
+                    stores_to_send.push(&store_name);
+                }
+            }
+        }
+
+        for store_name in stores_to_send {
+            if let Err(err) = glean.send_ping_by_name(store_name, false) {
+                log::error!(
+                    "Got more than {} events, but could not send {} ping: {}",
+                    glean.get_max_events(),
+                    store_name,
+                    err
+                );
+            }
+        }
+    }
+
+    /// Writes an event to a single store on disk.
+    ///
+    /// This assumes that the write lock in `self.event_stores` is already held.
+    ///
+    /// # Arguments
+    ///
+    /// - `store_name`: The name of the store.
+    /// - `event_json`: The event content, as a single-line JSON-encoded string.
+    fn write_event_to_disk(&self, store_name: &str, event_json: &str) {
+        if let Err(err) = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(self.path.join(store_name))
+            .and_then(|mut file| writeln!(file, "{}", event_json))
+        {
+            log::error!("Error writing event to store {}: {}", store_name, err);
+        }
+    }
+
+    /// Get a snapshot of the stored event data as a JsonValue.
+    ///
+    /// # Arguments
+    ///
+    /// - `store_name` The name of the desired store.
+    /// - `clear_store` Whether to clear the store afterward.
+    ///
+    /// # Returns
+    ///
+    /// The JsonValue as an array of events, if any.
+    pub fn snapshot_as_json(&self, store_name: &str, clear_store: bool) -> Option<JsonValue> {
+        let result = {
+            let db = self.event_stores.read().unwrap();
+            db.get(&store_name.to_string()).and_then(|store| {
+                if !store.is_empty() {
+                    let first_timestamp = store[0].timestamp;
+                    Some(json!(store
```

Comment:
Weird formatting here. Did rustfmt run?

Reply:
Yep -- `rustfmt` is insisting on this :man_shrugging:

Reply:
Yikes.